include/clap/factory/draft/preset-discovery.h

```diff
@@ -0,0 +1,293 @@
+
+   // Retrives the path to a watch file.
+   // Whenever the given file is "touched", then the indexer shall invalidate all the data.
+   bool(CLAP_ABI *invalidation_watch_file)(const struct clap_preset_discovery_provider *provider,
+                                           char    *watch_file_path,
+                                           uint32_t watch_file_path_capacity);
+
+   // reads metadata from the given file and passes them to the metadata receiver
+   bool(CLAP_ABI *get_metadata)(const struct clap_preset_discovery_provider     *provider,
+                                const char                                      *uri,
+                                const clap_preset_discovery_metadata_receiver_t *metadata_receiver);
+} clap_preset_discovery_provider_t;
+
+typedef struct clap_preset_discovery_indexer {
+   clap_version_t clap_version; // initialized to CLAP_VERSION
+   const char    *name;
+   const char    *vendor;
+   const char    *version;
+
+   // Registers a preset filetype
+   void(CLAP_ABI *register_filetype)(const struct clap_preset_discovery_indexer *indexer,
+                                     const clap_preset_discovery_filetype_t     *filetype);
+
+   // Registers a preset location
+   void(CLAP_ABI *register_location)(const struct clap_preset_discovery_indexer *indexer,
+                                     const clap_preset_discovery_location_t     *location);
+
+   // Registers a preset collection
+   void(CLAP_ABI *register_collection)(const struct clap_preset_discovery_indexer *indexer,
+                                       const clap_preset_discovery_collection_t   *collection);
+} clap_preset_indexer_t;
+
+// Every methods in this factory must be thread-safe.
+// It is encourraged to perform preset indexing in background threads, maybe even in background
+// process.
+typedef struct clap_preset_discovery_factory {
+   // Get the number of preset providers available.
+   // [thread-safe]
+   uint32_t(CLAP_ABI *count)(const struct clap_preset_discovery_factory *factory);
+
+   // Retrieves a preset provider descriptor by its index.
+   // Returns null in case of error.
+   // The descriptor must not be freed.
+   // [thread-safe]
+   const clap_preset_discovery_provider_descriptor_t *(CLAP_ABI *get_descriptor)(
+      const struct clap_preset_discovery_factory *factory, uint32_t index);
+
+   // Create a preset provider by its id.
+   // The returned pointer must be freed by calling preset_provider->destroy(preset_provider);
+   // The preset provider is not allowed to use the indexer callbacks in the create method.
+   // Returns null in case of error.
+   // [thread-safe]
+   const clap_preset_discovery_provider_t *(CLAP_ABI *create)(
+      const struct clap_preset_discovery_factory *factory,
+      const clap_preset_indexer_t                *indexer,
+      const char                                 *provider_id);
```

Comment:
Would it make sense to have a global invalidation file? That could be set to, e.g., the plugin's config file or some preset-specific config file. If that changes, then that would cause the host to rescan for added or removed preset providers. Right now there's no way to add or remove a plugin location, and this would enable that.

Reply:
I think it'd would like for the plugin factory.

Reply:
What do you mean? Or do you mean that we could use the plugin invalidation factory for this in some way?

Reply:
It is the exact same pattern, so maybe it could be generalized.